src/indicators/drought/arc2_precipitation.py

```diff
@@ -0,0 +1,710 @@
+            only looked for on or after that date.
+        """
+        if master:
+            self._download(self.date_min, self.date_max, master)
+        else:
+            # load master data and find all dates covered
+            # compare to min/max and then download missing
+            mr = self.load_raw_data()
+
+            # drop missing data if requested
+            if replace_missing:
+                if isinstance(replace_missing, str):
+                    replace_date = datetime.fromisoformat(replace_missing)
+                else:
+                    replace_date = datetime.combine(
+                        self.date_min, datetime.min.time()
+                    )
+
+                t_subset = mr.indexes["T"] < replace_date
+                val_subset = np.max(
+                    np.max(mr.values != -999, array=1), array=1
+                )
+
+                # keeps rows if date is less than specified or
+                # value is different than -999 (missing)
+                mr = mr.loc[t_subset | val_subset, :, :]
+
+            loaded_dates = mr.indexes[
+                "T"
+            ].to_datetimeindex() - pd.to_timedelta("12:00:00")
+            full_dates = pd.date_range(self.date_min, self.date_max)
+            needed_dates = full_dates.difference(loaded_dates)
+
+            if len(needed_dates) > 0:
+                date_ranges = self._group_date_ranges(needed_dates)
+                for dates in date_ranges:
+                    self._download(
+                        date_min_dl=dates[0],
+                        date_max_dl=dates[1],
+                        master=False,
+                    )
+
+                self._sort_raw_data()
+            else:
+                logger.info("No additional data needs downloading.")
+
+    def _group_date_ranges(self, dates) -> list:
+        """
+        Group range of dates into list of consecutive dates to pass to
+        `self._download()` for calling to the API. For each group, only
+        the min and max dates are returned rather than the full list
+        of dates in the group.
+
+        :param dates: Date range generated by `pd.date_range()`
+        """
+        date_ranges = []
+
+        for _, g in groupby(
+            enumerate(dates),
+            key=lambda x: x[0] - (x[1] - datetime(1970, 1, 1)).days,
+        ):
+            group = map(itemgetter(1), g)
+            group = list(map(pd.Timestamp, group))
+            if len(group) == 1:
+                group.append(group[0])
+            date_ranges.append((group[0], group[-1]))
+
+        return date_ranges
+
+    def _sort_raw_data(self):
+        """
+        Sort master file by time coordinates to ensure
+        correct ordering.
+        """
+        master_filepath = self._get_raw_filepath(master=True)
+        with xr.open_dataarray(master_filepath) as ds:
+            master = ds.load()
+
+        master.sortby("T").to_netcdf(master_filepath)
+
+
+class DrySpells(ARC2):
```

Comment:
Would it be an idea to move the dryspells to a separate file? And maybe don't make it dependent on ARC2? Such that other datasets such as CHIRPS could also be given as input (at first glance code doesn't seem ARC2 specific? )

Reply:
Yes, definitely the idea, will be part of refactor into toolbox!